### ISSUE TITLE
Emacs treesit experiments

### ISSUE THREAD

**Issue by dvzubarev** [meain/evil-textobj-tree-sitter#109]
Hi, I wanted to share my experiments with the Emacs builtin treesit library.
Recently, there were added some functions for creating arbitrary thing at points from the parse tree nodes.
I had an idea to implement text objects based on these things.
The basic idea is to search current thing at point.
Then extend or shrink its bounds according to current modifier (inner, outer, etc.).

There are pros and cons for this approach in comparison with the query based approach.
On the good side, It is more efficient approach, since it does not require to query the whole buffer with bunch of queries.
Its working by local traversing of the tree nodes, that are near the point.
So performance of this approach should not be affected by the size of the buffer.
Also, it allows more fine-grained control of text object ranges.
For example, for parameter outer, its possible to include surrounding spaces.
There are also possibilities to make fancier text objects.

The main downside is the lack of language support.
Since we cannot use community queries, we have to create settings for each language from the scratch.

I ended up crafting a package out of this experiments for my personal use - https://github.com/dvzubarev/evil-ts-obj.
Please feel free to check it out, it would be great to hear some feedback.

**Comment by meain** [comment]
This looks interesting. Relying on the named fields seems like a good idea(not sure about the support for every language though). FYI, there is a package which does something vaguely similar [foxfriday/evil-ts](https://github.com/foxfriday/evil-ts).

> It is more efficient approach, since it does not require to query the whole buffer with bunch of queries.

Hmm, that sound useful. That said, I've not really had run into any issues with perf even on relatively large buffers.

> Also, it allows more fine-grained control of text object ranges.
For example, for parameter outer, its possible to include surrounding spaces.

The `elisp-tree-sitter` variant lets you do similar things. The reason why we can't support this is due to missing functionality to get [matches](https://lists.gnu.org/archive/html/emacs-devel/2023-03/msg00500.html).

> The main downside is the lack of language support.

Yeah, my plan from the start was to just copy the "homework" done by neovim folks. It has not always been that smooth because of neovim using a lot of custom operators.

**Comment by dvzubarev** [comment]
> if you were to do parameter.inner, it would select the "b" as we select the very next one if we are not on one.

Yes, but this approach does not work if there are nested text objects. For example `print( call(a,| b, c) )` will select `call(a,| b, c)` as `parameter.inner`. In most cases, it is not what I want to do.

**Comment by meain** [comment]
Ahh, I see. That would definitely need custom handling.
